read move function types from normalized modules

gegaowp opened the pull request:
the move function types can be used on TS side to tell users the permissions txn requests for related objects, for example:
- read only
- read + write
- full access including transfer

Test via fullnode JSON RPC
```
curl --location --request POST http://127.0.0.1:9000 \
--header 'Content-Type: application/json' \
--data-raw '{ "jsonrpc":"2.0", "method":"sui_getMoveFunctionArgTypes","id":1, "params":["0x0000000000000000000000000000000000000002", "balance", "create_supply"]}'
{"jsonrpc":"2.0","result":["Pure"],"id":1}%                  

                                              
curl --location --request POST http://127.0.0.1:9000 \
--header 'Content-Type: application/json' \
--data-raw '{ "jsonrpc":"2.0", "method":"sui_getMoveFunctionArgTypes","id":1, "params":["0x0000000000000000000000000000000000000002", "coin", "destroy_zero"]}'
{"jsonrpc":"2.0","result":[{"Object":"ByValue"}],"id":1}%  
```

emmazzz commented on crates/sui-json-rpc/src/read_api.rs:
```diff
@@ -121,6 +123,52 @@ impl SuiRpcModule for ReadApi {
+                        module: _,
+                        name: _,
+                        type_arguments: _,
+                    } => "TRANSFER".into(),
```
Maybe use something like "FULL ACCESS" here? Since having a struct here in the parameter means the function has full access of the object and can read, write, transfer, destroy, etc.

tnowacki replied:
"BY VALUE" is also a term we use elsewhere.

But I think it is hard to capture the write terminology here:
- "TRANSFER" has issues since it might be destroyed 
- "FULL ACCESS" feels like it is somehow just read+write? or somehow doesn't really convey to me that it might be destroyed

emmazzz commented on crates/sui-json-rpc/src/read_api.rs:
```diff
@@ -121,6 +123,52 @@ impl SuiRpcModule for ReadApi {
+        package: ObjectID,
+        module: String,
+        function: String,
+    ) -> RpcResult<Vec<String>> {
```
Wondering why we use strings here to represent the tags instead of an enum?

gegaowp replied:
I was thinking this function will only be called from typescript side so the enum type will not help much, but I will do that.

tnowacki commented on crates/sui-types/src/move_package.rs:
```diff
@@ -63,6 +64,10 @@ impl MovePackage {
     pub fn disassemble(&self) -> SuiResult<BTreeMap<String, Value>> {
         disassemble_modules(self.module_map.values())
     }
+
+    pub fn normalize(&self) -> SuiResult<BTreeMap<String, Module>> {
```
nit, `normalized::Module` might read a bit better here

tnowacki replied:
I'm a bit worried about the creep of APIs like this attached `MovePackage`. There is no guarantee that a `MovePackage` value deserializes or passes the verifier, since the type is also used for failed transactions. (the one below is fine)

tnowacki commented on crates/sui-json-rpc/src/read_api.rs:
```diff
@@ -121,6 +123,52 @@ impl SuiRpcModule for ReadApi {
+                        type_arguments: _,
+                    } => "TRANSFER".into(),
+                    Type::Reference(_) => "READ".into(),
+                    Type::MutableReference(_) => "WRITE".into(),
```
Maybe "READ+WRITE"?

666lcz replied:
Should we just return `Reference`, `MutableReference`, and `Struct` and leave the client to interpret?

tnowacki replied:
I think that could also work. But I would probably suggest "by-value" or something similar instead of `Struct`

gegaowp commented:
thanks a lot for the quick comments. I updated the PR with changes:
- removed normalize() from MovePackage
- pass forward the type and let TS side to interpret the semantic meaning
- other nits.
ready for another review now, thanks!

tnowacki commented on crates/sui-json-rpc/src/read_api.rs:
```diff
@@ -121,6 +125,60 @@ impl SuiRpcModule for ReadApi {
+                        name: _,
+                        type_arguments: _,
+                    } => MoveFunctionArgType::Struct,
+                    Type::Vector(_) => MoveFunctionArgType::Vector,
```
If you are going to do this, why not just provide the full type? `Vector` or `TypeParameter` or `Struct` don't seem that helpful to me unless I also know the whole type?

If you want to capture the kind/intent of the argument here, I might suggest
```
Pure,
Object(ObjectValueKind),
```
where
```
ObjectValueKind {
  ByImmutableReference,
  ByMutableReference,
  ByValue,
}
```

gegaowp replied:
agreed that here Vector is probably not very useful and the suggested type should be enough for the wallet use-case, @666lcz mentioned that sometime types like U64 can be useful. Chris can you add some contexts here?

tnowacki replied:
I think you should have:
(A) full type information
(B) pure or object+kind
(C) nothing :p 

Having some weird mismatch of partial type information I think will be confusing as it will not line up with how Move will be looking at these values

gegaowp commented:
also ran 
```
cargo -q run --example generate-json-rpc-spec -- record
```
for some json rpc generated changes

gegaowp commented:
@tnowacki thanks for the detailed comment, I made some changes accordingly and also updated the test plan. thanks!

666lcz commented on crates/sui-json-rpc/src/read_api.rs:
```diff
@@ -121,6 +125,64 @@ impl SuiRpcModule for ReadApi {
+            Some(parameters) => Ok(parameters
+                .iter()
+                .map(|p| match p {
+                    Type::Bool => MoveFunctionArgType::Pure,
+                    Type::U8 => MoveFunctionArgType::Pure,
+                    Type::U64 => MoveFunctionArgType::Pure,
+                    Type::U128 => MoveFunctionArgType::Pure,
+                    Type::Address => MoveFunctionArgType::Pure,
+                    Type::Signer => MoveFunctionArgType::Pure,
+                    Type::Struct {
+                        address: _,
+                        module: _,
+                        name: _,
+                        type_arguments: _,
+                    } => MoveFunctionArgType::Object(ObjectValueKind::ByValue),
+                    Type::Vector(_) => MoveFunctionArgType::Pure,
+                    Type::TypeParameter(_) => MoveFunctionArgType::Pure,
+                    Type::Reference(_) => {
+                        MoveFunctionArgType::Object(ObjectValueKind::ByImmutableReference)
+                    }
+                    Type::MutableReference(_) => {
+                        MoveFunctionArgType::Object(ObjectValueKind::ByMutableReference)
+                    }
```
```suggestion
                    Type::Struct {
                        address: _,
                        module: _,
                        name: _,
                        type_arguments: _,
                    } => MoveFunctionArgType::Object(ObjectValueKind::ByValue),
                    Type::Reference(_) => {
                        MoveFunctionArgType::Object(ObjectValueKind::ByImmutableReference)
                    }
                    Type::MutableReference(_) => {
                        MoveFunctionArgType::Object(ObjectValueKind::ByMutableReference)
                    },
                    _ => MoveFunctionArgType::Pure,
```

666lcz commented on crates/sui-json-rpc/src/read_api.rs:
```diff
@@ -121,6 +125,64 @@ impl SuiRpcModule for ReadApi {
+                    }
+                })
+                .collect::<Vec<MoveFunctionArgType>>()),
+            None => Err(anyhow!("No parameters found for function {}", function)),
```
Can we test this case? Will a function like `create()` trigger an error?

666lcz commented on crates/sui-json-rpc/src/read_api.rs:
```diff
@@ -121,6 +125,64 @@ impl SuiRpcModule for ReadApi {
 
 #[async_trait]
 impl RpcFullNodeReadApiServer for FullNodeApi {
+    async fn get_move_function_arg_types(
```
Can you also add sample data generation for this endpoint: https://github.com/MystenLabs/sui/blob/bfb41f2d36d4cd4d080c91253f66583bb212655e/crates/sui-open-rpc/src/generate_json_rpc_spec.rs#L148 ?

gegaowp commented:
- generated move.json file
- tested for non-existing function
![Screen Shot 2022-08-08 at 12 31 26 PM](https://user-images.githubusercontent.com/106119108/183467771-e3b389cf-6be0-4053-b449-cd2695b838d5.png)
